i18n-helpers/src/bin/mdbook-i18n.rs

```diff
@@ -0,0 +1,107 @@
+use mdbook::renderer::RenderContext;
+use serde::Deserialize;
+use std::collections::BTreeMap;
+use std::io;
+
+/// Parameters for the i18n renderer.
+///
+/// They are set in the `output.i18n` section of the book's `book.toml` file.
+#[derive(Deserialize)]
+struct I18nConfiguration {
+    /// A map of language codes to language names.
+    ///
+    /// ## Example
+    ///
+    /// ```toml
+    /// [output.i18n.languages]
+    /// "en" = "English"
+    /// "es" = "Spanish (Español)"
+    /// "ko" = "Korean (한국어)"
+    /// "pt-BR" = "Brazilian Portuguese (Português do Brasil)"
+    /// ```
+    #[serde(default)]
+    languages: BTreeMap<String, String>,
+    /// Default language code. It will not be translated.
+    default_language: Option<String>,
+
+    /// Whether to translate all languages or just the selected language, defaults to false.
```

Comment:
The language set in the env variable `MDBOOK_BOOK__LANGUAGE`

Reply:
Ah, okay... Setting `MDBOOK_BOOK__LANGUAGE` is just another way of setting the `language` field in `book.toml`. So it's not strictly "selected", it's just what the config happens to be when the `mdbook-gettext` preprocessor is invoked.

The current system has the elegant approach that you can set `language` to what you want and it will then translate the text on the fly to this language — because it's a preprocessor which is independent of the renderer!

The new renderer is a bit different: it takes a list of languages from `book.toml` and renders the book for each of them. If the user only want to render a single language, the simplest way would be for the user to remove the other languages from the list in `book.toml`!

The users could do this on the command line with a suitable `MDBOOK_OUTPUT__I18n__LANGUAGES` setting (I think). This would be better since it uses less infrastructure: all you need to know is that the `output.i18n.languages` key controls the list of languages to generate for. If it has 1 item, then you get a single language, if it has $n$ items, then you get $n$ languages.

Reply:
The conversion is documented here: https://rust-lang.github.io/mdBook/format/configuration/environment-variables.html.

Reply:
The reason for this variable is that if you do  `MDBOOK_BOOK__LANGUAGE=fr mdbook build` with `translate_all_languages` to false, it will just translate `fr` otherwise, it will translate all languages. We can skip that, it's just for convenience.

Reply:
> The reason for this variable is that if you do `MDBOOK_BOOK__LANGUAGE=fr mdbook build` with `translate_all_languages` to false, it will just translate `fr` otherwise, it will translate all languages. We can skip that, it's just for convenience.

Can we unify this behavior? We stop setting `MDBOOK_BOOK__LANGUAGE` and instead drive the behavior of the backend using _solely_ a new `output.i18n.languages` field. It should behave like this:

- The `book.toml` file checked in has all languages for the book.
- To build fewer languages, people can remove the unwanted languages from `book.toml`. Alternativley, they can use the environment variable support and run

  ```sh
  MDBOOK_OUTPUT__I18N__LANGUAGES='["fr"]' mdbook build
  ```

  to restrict the languages to just French.

Why have a list in `book.toml` instead of just using `po/*.po`? Because there could be extra meta data compared to what we find in the PO files. I'm thinking of things like `english_name = "French", name = "Français"` and maybe even `available_in_language_picker = true`.

Does that sounds like a good idea? So instead of a "all languages or a single language" based on `MDBOOK_BOOK__LANGUAGE` (which doesn't make a ton of sense for a multi-language book), we have a more flexible system that people can override on the command line.

Reply:
> What you're suggesting would mean intercepting everything right at the beginning,

I'm not actually suggesting that — I was only suggesting changing the logic for which subset of the languages to build.